flink-connector-mysql-cdc/src/test/java/com/alibaba/ververica/cdc/connectors/mysql/MySqlTestUtils.java

```diff
@@ -0,0 +1,330 @@
+                        .username(database.getUsername())
+                        .password(database.getPassword())
+                        .deserializer(new ForwardDeserializeSchema())
+                        .debeziumProperties(createDebeziumProperties(useLegacyImplementation))
+                        .build();
+        final TestingListState<byte[]> offsetState = new TestingListState<>();
+        final TestingListState<String> historyState = new TestingListState<>();
+
+        // ---------------------------------------------------------------------------
+        // Step-1: start source
+        // ---------------------------------------------------------------------------
+        TestSourceContext<SourceRecord> sourceContext = new TestSourceContext<>();
+        setupSource(source, false, offsetState, historyState, true, 0, 1);
+        final CheckedThread runThread =
+                new CheckedThread() {
+                    @Override
+                    public void go() throws Exception {
+                        source.run(sourceContext);
+                    }
+                };
+        runThread.start();
+
+        drain(sourceContext, expectedRecordCount);
+
+        // ---------------------------------------------------------------------------
+        // Step-2: trigger checkpoint-1 after snapshot finished
+        // ---------------------------------------------------------------------------
+        synchronized (sourceContext.getCheckpointLock()) {
+            // trigger checkpoint-1
+            source.snapshotState(new StateSnapshotContextSynchronousImpl(101, 101));
+        }
+
+        assertEquals(1, offsetState.list.size());
+        String state = new String(offsetState.list.get(0), StandardCharsets.UTF_8);
+        String offsetFile = JsonPath.read(state, "$.sourceOffset.file");
+        int offsetPos = JsonPath.read(state, "$.sourceOffset.pos");
+
+        source.cancel();
+        source.close();
+        runThread.sync();
+
+        return Tuple2.of(offsetFile, offsetPos);
+    }
+
+    public static MySqlSource.Builder<SourceRecord> basicSourceBuilder(
+            UniqueDatabase database, boolean useLegacyImplementation) {
+        Properties debeziumProps = createDebeziumProperties(useLegacyImplementation);
+        return MySqlSource.<SourceRecord>builder()
+                .hostname(database.getHost())
+                .port(database.getDatabasePort())
+                .databaseList(database.getDatabaseName())
+                .tableList(
+                        database.getDatabaseName() + "." + "products") // monitor table "products"
+                .username(database.getUsername())
+                .password(database.getPassword())
+                .deserializer(new ForwardDeserializeSchema())
+                .debeziumProperties(debeziumProps);
+    }
+
+    static <T> void setupSource(DebeziumSourceFunction<T> source) throws Exception {
```

Comment:
I don't like package privilege for `Utils` class, maybe `public` is better, HDYT?